uncertainties/ucombo.py

```diff
@@ -0,0 +1,134 @@
+from __future__ import annotations
+
+from collections import defaultdict
+from math import sqrt
+from numbers import Real
+from typing import Optional, Tuple, Union
+import uuid
+
+
+class UAtom:
+    __slots__ = ["uuid", "tag", "_hash"]
+
+    def __init__(self: UAtom, tag: Optional[str] = None):
+        self.tag = tag
+        self.uuid: uuid.UUID = uuid.uuid4()
+        self._hash = hash(self.uuid)
+
+    def __eq__(self: UAtom, other: UAtom) -> bool:
+        return hash(self) == hash(other)
+
+    def __hash__(self: UAtom) -> int:
+        return self._hash
+
+    def __lt__(self: UAtom, other: UAtom) -> bool:
+        return self.uuid < other.uuid
```

Comment:
codecov gives a load of warnings for UAtom and UCombo. Would bit be worth having tests to show methods like eq, lt behave as expected?

it's also a bit suprising some of these lines (eg for UCombo's eq) don't get hit by any of the existing tests

Reply:
Thanks for taking a look. I haven't looked at codecov yet. The story for why `UCombo`s `__eq__` doesn't get hit is worth sharing.

First off, before this PR and as of this comment, `UFloat.__eq__` does not check for equality of the nominal value and equality of the uncertainty as I would have naively implemented. Rather, it looks at the difference between `self` and `other` and checks that the resulting `UFloat` has zero nominal value and zero std_dev. Because subtraction is supported between `UFloat` and `float`, this allows `__eq__` to easily support comparison between `UFloat` and `float`. In going through the tests I've learned that, for better or for worse, `uncertainties` has definitely chosen the convention that a `UFloat` with a `std_dev` of zero should behave like a regular `float`.

I am still using this difference calculation for `UFloat.__eq__`. It would be more natural to me to have `__eq__` be something like
```
def __eq__(self, other):
    if isinstance(other, UFloat):
        return (self.nominal_value == other.nominal_value and self.uncertainty = other.uncertainty)
    else:
        if self.std_dev != 0:
            return False
        else:
            return self.nominal_value == other
```
But currently the code implementing `__eq__` is defined in `ops.py` and gets monkey patched onto `UFloat`. I can't import `UFloat` into this function because it would create circular imports between `ops.py` and `core.py`.

Anyways, I can look into disentangling this.

________________
Note though, that I tried something like the above code for a second and found out that the current version of `UCombo.__eq__` was broken because I didn't know `sorted` returned a list! Easy fix, but emphasizes the importance of testing.

Note that I made the decision that `UCombo` `__eq__` and `__hash__` use the `expanded` dictionary for comparison. Thuis means that two `UCombo`s with different `ucombo_tuple`s could compare equal. For `UFloat` I think it is clear that we want two `UFloat`s to compare based on the expanded dictionary from the uncertainty. For `UCombo` it wasn't as obvious to me. For example, if `dx` is a `UAtom` then it would be possible for there to be a nested `UCombo` storing something like
```
dx + (-dx + (dx + (-dx + (dx + (-dx + ...)))))
```
Which ultimately is equal to 0. But the former might take a longer time to compute. 

Actually I think I need to add a special case for 0 so that the presence of a `UAtom` in one `UCombo` expanded dict but not the other doesn't matter if the weight is equal to 0.

In any case, does this make sense to allow two `UCombo` to compare the same even if they have different `ucombo_tuple`? Or should that distinction be made apparent in `__eq__`  and `__hash__` (but `UFloat` will still compare based on `expanded` dict)?